In [1]:
import pandas as pd

In [2]:
df = pd.read_parquet('../data/yellow_tripdata_2023-01.parquet')

In [3]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,2,2023-01-01 00:32:10,2023-01-01 00:40:36,1.0,0.97,1.0,N,161,141,2,9.3,1.00,0.5,0.00,0.0,1.0,14.30,2.5,0.00
1,2,2023-01-01 00:55:08,2023-01-01 01:01:27,1.0,1.10,1.0,N,43,237,1,7.9,1.00,0.5,4.00,0.0,1.0,16.90,2.5,0.00
2,2,2023-01-01 00:25:04,2023-01-01 00:37:49,1.0,2.51,1.0,N,48,238,1,14.9,1.00,0.5,15.00,0.0,1.0,34.90,2.5,0.00
3,1,2023-01-01 00:03:48,2023-01-01 00:13:25,0.0,1.90,1.0,N,138,7,1,12.1,7.25,0.5,0.00,0.0,1.0,20.85,0.0,1.25
4,2,2023-01-01 00:10:29,2023-01-01 00:21:19,1.0,1.43,1.0,N,107,79,1,11.4,1.00,0.5,3.28,0.0,1.0,19.68,2.5,0.00


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     str           
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            float64   

In [5]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
count,3.066766e+06,3066766,3066766,2.995023e+06,3.066766e+06,2.995023e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,3.066766e+06,2.995023e+06,2.995023e+06
mean,1.730215e+00,2023-01-17 00:22:26.288164,2023-01-17 00:38:06.427874,1.362532e+00,3.847342e+00,1.497440e+00,1.663980e+02,1.643926e+02,1.194483e+00,1.836707e+01,1.537842e+00,4.882900e-01,3.367941e+00,5.184907e-01,9.820847e-01,2.702038e+01,2.274231e+00,1.074086e-01
min,1.000000e+00,2008-12-31 23:01:42,2009-01-01 14:29:11,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,-9.000000e+02,-7.500000e+00,-5.000000e-01,-9.622000e+01,-6.500000e+01,-1.000000e+00,-7.510000e+02,-2.500000e+00,-1.250000e+00
25%,1.000000e+00,2023-01-09 16:21:57.250000,2023-01-09 16:37:06,1.000000e+00,1.060000e+00,1.000000e+00,1.320000e+02,1.140000e+02,1.000000e+00,8.600000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,1.000000e+00,1.540000e+01,2.500000e+00,0.000000e+00
50%,2.000000e+00,2023-01-17 08:42:29.500000,2023-01-17 08:58:30.500000,1.000000e+00,1.800000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,1.280000e+01,1.000000e+00,5.000000e-01,2.720000e+00,0.000000e+00,1.000000e+00,2.016000e+01,2.500000e+00,0.000000e+00
75%,2.000000e+00,2023-01-24 16:26:27,2023-01-24 16:42:49,1.000000e+00,3.330000e+00,1.000000e+00,2.340000e+02,2.340000e+02,1.000000e+00,2.050000e+01,2.500000e+00,5.000000e-01,4.200000e+00,0.000000e+00,1.000000e+00,2.870000e+01,2.500000e+00,0.000000e+00
max,2.000000e+00,2023-02-01 00:56:53,2023-02-02 09:28:47,9.000000e+00,2.589281e+05,9.900000e+01,2.650000e+02,2.650000e+02,4.000000e+00,1.160100e+03,1.250000e+01,5.316000e+01,3.808000e+02,1.969900e+02,1.000000e+00,1.169400e+03,2.500000e+00,1.250000e+00
std,4.438480e-01,NaN,NaN,8.961200e-01,2.495838e+02,6.474767e+00,6.424413e+01,6.994368e+01,5.294762e-01,1.780782e+01,1.789592e+00,1.034641e-01,3.826759e+00,2.017579e+00,1.833529e-01,2.216359e+01,7.718454e-01,3.556511e-01


In [9]:
df.shape   

(3066766, 19)

In [10]:
df.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge', 'airport_fee'],
      dtype='str')

In [11]:
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag                  str
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
airport_fee                     float64
dtype: object

In [12]:
df.index

RangeIndex(start=0, stop=3066766, step=1)

1. VendorID : Mã số của nhà cung cấp công nghệ đã ghi nhận chuyến đi. 1 = Creative Mobile Technologies, 2 = Curb Mobility, 6 = Myle Technologies Inc, 7 = Helix.

2.tpep_pickup_datetime: Ngày và giờ đồng hồ tính tiền được kích hoạt (bắt đầu chuyến đi).

3. tpep_dropoff_datetime: Ngày và giờ đồng hồ tính tiền ngừng hoạt động (kết thúc chuyến đi).

4. passenger_count: Số lượng hành khách trên xe (do tài xế khai báo). Lưu ý: có thể chứa giá trị 0 (lỗi nhập liệu) hoặc các giá trị ngoại lai như 9.

5. trip_distance: Quãng đường đã đi được tính bằng dặm, do đồng hồ tính tiền ghi nhận.

6. RatecodeID: Mã loại giá cước cuối cùng của chuyến đi: 1 = Standard, 2 = JFK, 3 = Newark, 4 = Nassau/Westchester, 5 = Thương lượng, 6 = Đi chung, 99 = Không rõ.

7. store_and_fwd_flag: Cờ "lưu trữ và chuyển tiếp": "Y" nếu dữ liệu được lưu trong bộ nhớ xe trước khi gửi về máy chủ (do mất kết nối), "N" nếu gửi trực tiếp.

8. PULocationID: Mã số khu vực taxi (Taxi Zone) nơi khách lên xe.

9. DOLocationID: Mã số khu vực taxi nơi khách xuống xe.

10. payment_type: Mã phương thức thanh toán: 0 = Flex Fare, 1 = Credit Card, 2 = Cash, 3 = No charge, 4 = Dispute, 5 = Unknown, 6 = Voided trip.

11. fare_amount: Số tiền cước phí dựa trên thời gian và quãng đường tính bởi đồng hồ.

12. extra: Các khoản phụ thu và phụ phí linh tinh.

13. mta_tax: Thuế MTA (Cơ quan Giao thông Đô thị) được tự động kích hoạt dựa trên mức cước đang sử dụng.

14. tip_amount: Số tiền tip. Chỉ tự động ghi nhận đối với tip qua thẻ tín dụng. Tip tiền mặt không được bao gồm.

15. tolls_amount: Tổng số tiền phí cầu đường đã trả trong chuyến đi.

16. improvement_surcharge: Phụ phí cải thiện taxi ($0.30 cho mỗi chuyến) bắt đầu từ năm 2015 để hỗ trợ cải thiện khả năng tiếp cận.

17. total_amount: Tổng số tiền khách trả, không bao gồm tip tiền mặt.

18. congestion_surcharge: Phụ phí ùn tắc giao thông của bang New York.

19. airport_fee: Phí sân bay, chỉ áp dụng khi đón khách tại LaGuardia và JFK.

- Dữ liệu có tổng cộng bao nhiêu dòng và bao nhiêu cột?

3066766 dòng, 19 cột


- Có những cột nào liên quan đến thời gian?

tpep_pickup_datetime và tpep_dropoff_datetime


- Có những cột nào liên quan đến vị trí?

PULocationID và DOLocationID


- Có những cột nào liên quan đến tiền bạc?

fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge, total_amount, congestion_surcharge, airport_fee


- Có cột nào có vẻ như chứa giá trị bị thiếu (null) không? (gợi ý: dùng .info())
passenger_count, RatecodeID, store_and_fwd_flag, congestion_surcharge, 
airport_fee


- Giá trị passenger_count trung bình là bao nhiêu?

cột số khách
1.362532e+00: tức là trung bình khoảng 1.36 hành khách 


- total_amount trung bình là bao nhiêu?

cột tổng số tiền khách trả
2.702038e+01 = 2.702038 × 10^1 = 27.02038.
Kết luận: Tổng số tiền trung bình cho một chuyến taxi là khoảng 27.02 USD.

In [ ]:
df.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          71743
trip_distance                0
RatecodeID               71743
store_and_fwd_flag       71743
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     71743
airport_fee              71743
dtype: int64

In [4]:
no_passenger = df[df['passenger_count']==0].shape[0]
print(f"So chuyen khong co khach: {no_passenger}")

So chuyen khong co khach: 51164


In [5]:
max_row = df.loc[df['total_amount'].idxmax()]
print("Chuyen co tong tien cao nhat:")
print(max_row[['tpep_pickup_datetime','trip_distance','total_amount']])

Chuyen co tong tien cao nhat:
tpep_pickup_datetime    2023-01-24 12:43:44
trip_distance                        177.88
total_amount                         1169.4
Name: 2222612, dtype: object


In [9]:
# Lọc các chuyến đón ở Times Square
ts_pickups = df[df['PULocationID'] == 230]
avg_distance_ts = ts_pickups['trip_distance'].mean()
print(f"Quãng đường trung bình từ Times Square: {avg_distance_ts:.2f} dặm")

Quãng đường trung bình từ Times Square: 3.37 dặm


In [17]:
df['trip_duration_minutes'] = (df['tpep_dropoff_datetime']-df['tpep_pickup_datetime']).dt.total_seconds()/60

valid_trips = df[(df['trip_duration_minutes']>0)&(df['trip_duration_minutes']<180)]

print(f"Trung binh(mean): {valid_trips['trip_duration_minutes'].mean():.1f} phut")
print(f"Trung vi(median): {valid_trips['trip_duration_minutes'].median():.1f} phut")

Trung binh(mean): 14.5 phut
Trung vi(median): 11.5 phut


In [5]:
# Tạo cột giờ và ngày trong tuần
df['pickup_hour'] = df['tpep_pickup_datetime'].dt.hour
df['pickup_dayofweek'] = df['tpep_pickup_datetime'].dt.dayofweek  # 0=Thứ Hai, 6=Chủ Nhật

# Số chuyến theo giờ
hourly_counts = df.groupby('pickup_hour').size().sort_values(ascending=False)
print("Top 5 khung giờ đông nhất:")
print(hourly_counts.head(5))

# Bonus: Số chuyến theo giờ và ngày trong tuần (pivot table)
day_hour_pivot = df.pivot_table(index='pickup_dayofweek', columns='pickup_hour', aggfunc='size')
print("\nHeatmap sơ bộ (hàng=thứ, cột=giờ):")
print(day_hour_pivot)

Top 5 khung giờ đông nhất:
pickup_hour
18    215889
17    209493
15    196424
16    195977
19    192801
dtype: int64

Heatmap sơ bộ (hàng=thứ, cột=giờ):
pickup_hour          0      1      2      3     4     5     6      7      8   \
pickup_dayofweek                                                               
0                  7130   3682   2008   1537  1552  2646  7102  13244  17830   
1                  5945   2779   1521   1031  1225  3109  9226  19194  25336   
2                  5878   2634   1387    973  1062  2472  7041  14844  19963   
3                  7284   3497   2050   1289  1290  2390  7374  15618  20725   
4                 10543   5737   3051   1822  1621  2577  6202  13114  16836   
5                 20965  17250  12786   7907  3944  1899  3128   5299   8324   
6                 27224  24220  19237  12879  7141  2918  3787   5564   7851   

pickup_hour          9   ...     14     15     16     17     18     19     20  \
pickup_dayofweek         ...                 

In [6]:
# Tạo cột tip_percent
df['tip_percent'] = (df['tip_amount'] / df['fare_amount']) * 100

# Chỉ lấy thanh toán bằng thẻ để so sánh tip
card_trips = df[(df['payment_type'] == 1) & (df['fare_amount'] > 0)]

# Phân loại giờ cao điểm sáng
card_trips['is_rush_hour'] = card_trips['pickup_hour'].isin([7, 8, 9])

rush_tip = card_trips[card_trips['is_rush_hour']]['tip_percent'].median()
offpeak_tip = card_trips[~card_trips['is_rush_hour']]['tip_percent'].median()

print(f"Tip trung vị giờ cao điểm sáng (7-9AM): {rush_tip:.1f}%")
print(f"Tip trung vị giờ thấp điểm: {offpeak_tip:.1f}%")

Tip trung vị giờ cao điểm sáng (7-9AM): 25.4%
Tip trung vị giờ thấp điểm: 26.2%


In [3]:
df.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          71743
trip_distance                0
RatecodeID               71743
store_and_fwd_flag       71743
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     71743
airport_fee              71743
dtype: int64

Có 5 cột có dữ liệu null: 71743 giá trị
1. passenger_count: Số lượng hành khách trên xe
2. RatecodeID: Mã loại giá cước cuối cùng của chuyến đi
3. store_and_fwd_flag: Cờ "lưu trữ và chuyển tiếp"
4. congestion_surcharge: Phụ phí ùn tắc giao thông của bang New York.
5. airport_fee: Phí sân bay, chỉ áp dụng khi đón khách tại LaGuardia và JFK.

In [4]:
# Xóa toàn bộ các dòng có giá trị null ở cột 'passenger_count'
df_clean = df.dropna(subset=['passenger_count']).copy()

In [5]:
# Tìm giá trị xuất hiện nhiều nhất của cột 'RatecodeID'
mode_ratecode = df_clean['RatecodeID'].mode()[0]
# Thay thế các giá trị null bằng giá trị mode vừa tìm được
df_clean['RatecodeID'] = df_clean['RatecodeID'].fillna(mode_ratecode)

In [6]:
# Xóa cột 'store_and_fwd_flag' khỏi DataFrame
df_clean = df_clean.drop(columns=['store_and_fwd_flag'])

In [7]:
# Thay thế giá trị null bằng 0
df_clean['congestion_surcharge'] = df_clean['congestion_surcharge'].fillna(0)
df_clean['airport_fee'] = df_clean['airport_fee'].fillna(0)

In [8]:
df_clean.isnull().sum()

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
airport_fee              0
dtype: int64

In [9]:
df_clean.shape

(2995023, 18)

In [10]:
df.shape

(3066766, 19)